src/go/k8s/controllers/redpanda/cluster_controller.go

```diff
@@ -1171,3 +1125,353 @@ func isRedpandaClusterVersionManaged(
+	cluster    *vectorizedv1alpha1.Cluster
+	items      map[string]resources.Resource
+}
+
+const (
+	bootstrapService        = "BootstrapService"
+	clusterRole             = "ClusterRole"
+	clusterRoleBinding      = "ClusterRoleBinding"
+	clusterService          = "ClusterPorts"
+	configMap               = "ConfigMap"
+	headlessService         = "HeadlessService"
+	ingress                 = "Ingress"
+	nodeportService         = "NodeportService"
+	pki                     = "PKI"
+	podDisruptionBudget     = "PodDisruptionBudget"
+	proxySuperuser          = "ProxySuperuser"
+	schemaRegistrySuperuser = "SchemaRegistrySuperuser"
+	secret                  = "Secret"
+	serviceAccount          = "ServiceAccount"
+	statefulSet             = "StatefulSet"
+)
+
+func newAttachedResources(ctx context.Context, r *ClusterReconciler, log logr.Logger, cluster *vectorizedv1alpha1.Cluster) *attachedResources {
+	return &attachedResources{
+		ctx:        ctx,
+		reconciler: r,
+		log:        log,
+		cluster:    cluster,
+	}
+}
+
+func (a *attachedResources) Ensure() (ctrl.Result, error) {
+	result := ctrl.Result{}
+	var errs error
+	for _, resource := range a.items {
+		err := resource.Ensure(a.ctx)
+		var e *resources.RequeueAfterError
+		if errors.As(err, &e) {
+			a.log.Info(e.Error())
+			if result.RequeueAfter < e.RequeueAfter {
+				result = ctrl.Result{RequeueAfter: e.RequeueAfter}
+			}
+		} else if err != nil {
+			a.log.Error(err, "Failed to reconcile resource")
+			errs = errors.Join(errs, err)
+		}
+	}
+	return result, errs
+}
+
+func (a *attachedResources) bootstrapService() {
+	// if already initialized, exit immediately
+	if _, ok := a.items[bootstrapService]; ok {
+		return
+	}
+	redpandaPorts := networking.NewRedpandaPorts(a.cluster)
+	loadbalancerPorts := collectLBPorts(redpandaPorts)
+	a.items[bootstrapService] = resources.NewLoadBalancerService(a.reconciler.Client, a.cluster, a.reconciler.Scheme, loadbalancerPorts, true, a.log)
+}
+
+func (a *attachedResources) getBootstrapService() *resources.LoadBalancerServiceResource {
+	a.bootstrapService()
+	return a.items[bootstrapService].(*resources.LoadBalancerServiceResource)
+}
+
+func (a *attachedResources) clusterRole() {
+	// if already initialized, exit immediately
+	if _, ok := a.items[clusterRole]; ok {
+		return
+	}
+	a.items[clusterRole] = resources.NewClusterRole(a.reconciler.Client, a.cluster, a.reconciler.Scheme, a.log)
+}
+
+func (a *attachedResources) clusterRoleBinding() {
+	// if already initialized, exit immediately
+	if _, ok := a.items[clusterRoleBinding]; ok {
+		return
+	}
+	a.items[clusterRole] = resources.NewClusterRole(a.reconciler.Client, a.cluster, a.reconciler.Scheme, a.log)
+}
+
+func (a *attachedResources) getClusterRoleBinding() *resources.ClusterRoleBindingResource {
+	a.clusterRoleBinding()
+	return a.items[clusterRoleBinding].(*resources.ClusterRoleBindingResource)
+}
+
+func (a *attachedResources) clusterService() {
+	// if already initialized, exit immediately
```

Comment:
Why not all function has such checks?

Reply:
We only need it for functions that create new instances of structs.